java/ql/lib/semmle/code/java/dataflow/TypeFlow.qll

```diff
@@ -146,27 +156,165 @@ private predicate joinStep(TypeFlowNode n1, TypeFlowNode n2) {
+    not sccEdgeWithMax(n, n2, n)
+    or
+    sccEdgeWithMax(n2, n, n2) and
+    not sccEdgeWithMax(n2, n, n)
+  )
+}
+
+private predicate localMax(TypeFlowNode m) {
+  sccEdgeWithMax(_, _, m) and
+  not hasLargerNeighbor(m)
 }
 
-private predicate joinStepRank(int r, TypeFlowNode n1, TypeFlowNode n2) {
-  exists(int r1 |
-    joinStepRank1(r1, n1, n2) and
-    joinStepRank2(r, r1, n2)
+private predicate sccReprFromLocalMax(TypeFlowNode scc) {
+  exists(TypeFlowNode m |
+    localMax(m) and
+    scc =
+      max(TypeFlowNode n2 |
+        sccEdge+(m, n2) and localMax(n2)
+      |
+        n2
+        order by
+          n2.getLocation().getStartLine(), n2.getLocation().getStartColumn(), getNodeKind(n2)
+      )
   )
 }
 
-private int lastRank(TypeFlowNode n) { result = max(int r | joinStepRank(r, _, n)) }
+private predicate sccRepr(TypeFlowNode n, TypeFlowNode scc) {
+  sccEdge+(n, scc) and sccReprFromLocalMax(scc)
+}
+
+private predicate sccJoinStep(TypeFlowNode n, TypeFlowNode scc) {
+  exists(TypeFlowNode mid |
+    joinStep(n, mid) and
+    sccRepr(mid, scc) and
+    not sccRepr(n, scc)
+  )
+}
+
+private signature predicate edgeSig(TypeFlowNode n1, TypeFlowNode n2);
+
+private signature module RankedEdge {
+  predicate edgeRank(int r, TypeFlowNode n1, TypeFlowNode n2);
+
+  int lastRank(TypeFlowNode n);
+}
+
+private module RankEdge<edgeSig/2 edge> implements RankedEdge {
+  private predicate edgeRank1(int r, TypeFlowNode n1, TypeFlowNode n2) {
+    n1 =
+      rank[r](TypeFlowNode n |
+        edge(n, n2)
+      |
+        n order by n.getLocation().getStartLine(), n.getLocation().getStartColumn()
+      )
+  }
+
+  private predicate edgeRank2(int r2, int r1, TypeFlowNode n) {
+    r1 = rank[r2](int r | edgeRank1(r, _, n) | r)
+  }
+
+  predicate edgeRank(int r, TypeFlowNode n1, TypeFlowNode n2) {
+    exists(int r1 |
+      edgeRank1(r1, n1, n2) and
+      edgeRank2(r, r1, n2)
+    )
+  }
+
+  int lastRank(TypeFlowNode n) { result = max(int r | edgeRank(r, _, n)) }
+}
+
+private signature module TypePropagation {
+  predicate candType(TypeFlowNode n, RefType t);
+
+  bindingset[t]
+  predicate supportsType(TypeFlowNode n, RefType t);
+}
+
+/** Implements recursion through `forall` by way of edge ranking. */
+private module ForAll<RankedEdge Edge, TypePropagation T> {
+  /**
+   * Holds if `t` is a bound that holds on one of the incoming edges to `n` and
+   * thus is a candidate bound for `n`.
+   */
+  pragma[nomagic]
+  private predicate candJoinType(TypeFlowNode n, RefType t) {
+    exists(TypeFlowNode mid |
+      T::candType(mid, t) and
+      Edge::edgeRank(_, mid, n)
+    )
+  }
+
+  /**
+   * Holds if `t` is a candidate bound for `n` that is also valid for data coming
+   * through the edges into `n` ranked from `1` to `r`.
+   */
+  private predicate flowJoin(int r, TypeFlowNode n, RefType t) {
+    (
+      r = 1 and candJoinType(n, t)
+      or
+      flowJoin(r - 1, n, t) and Edge::edgeRank(r, _, n)
```

Comment:
Semantically, yes. But the codegen will be worse. A `forex(vars | range | term)` currently desugars to `exists(vars | range | term) and forall(vars | range | term)` so that would give an additional recursive call (which we might or might not be lucky enough to have the optimiser deduplicate at the RA level after join-ordering, since the delta of a `forall` is equal to the delta of a `forex`).

Reply:
👍